src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java

```diff
@@ -192,4 +195,25 @@ private void notTheRealCauseOfTheTimeout() {
             }
         }
     }
+
+    @Test
+    public void threadGroupNotLeaked()
+            throws Throwable {
+
+        Set<ThreadGroup> groupsBeforeSet = new HashSet<ThreadGroup>();
+        ThreadGroup[] groupsBefore = new ThreadGroup[256];
+        for (int i = 0; i < currentThread().getThreadGroup().enumerate(groupsBefore); ++i) {
+            groupsBeforeSet.add(groupsBefore[i]);
+        }
+
+        evaluateWithWaitDuration(0);
+        
+        ThreadGroup[] groups = new ThreadGroup[256];
+        for (int i = 0; i < currentThread().getThreadGroup().enumerate(groups); ++i) {
+            if(!groupsBeforeSet.contains(groups[i]) && "FailOnTimeoutGroup".equals(groups[i].getName())) {
+                fail("A 'FailOnTimeoutGroup' thread group remains referenced after the test execution.");
+            }
+        }
+    }
```

Comment:
The readability of the test can be improved by extracting a method `subGroupsOfCurrentThread`

    @Test
    public void threadGroupNotLeaked()
            throws Throwable {

        Collection<ThreadGroup> groupsBeforeSet = subGroupsOfCurrentThread();

        evaluateWithWaitDuration(0);

        for (ThreadGroup group: subGroupsOfCurrentThread()) {
            if(!groupsBeforeSet.contains(group) && "FailOnTimeoutGroup".equals(group.getName())) {
                fail("A 'FailOnTimeoutGroup' thread group remains referenced after the test execution.");
            }
        }
    }

    private Collection<ThreadGroup> subGroupsOfCurrentThread() {
        ThreadGroup[] subGroups = new ThreadGroup[256];
        int numGroups = currentThread().getThreadGroup().enumerate(subGroups);
        return asList(subGroups).subList(0, numGroups);
    }